pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+func ParseRemoteReadRequestWithoutCosumingBody(req *http.Request) (url.Values, error) {
+	params := make(url.Values)
+
+	if req.Body == nil {
+		return params, nil
+	}
+
+	bodyBytes, err := util.ReadRequestBodyWithoutConsuming(req)
+
+	if req.Header.Get("Content-Encoding") == "snappy" {
+		bodyBytes, err = snappy.Decode(nil, bodyBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Parse the request data.
+	remoteReadRequest := prompb.ReadRequest{}
+	if err := proto.Unmarshal(bodyBytes, &remoteReadRequest); err != nil {
+		return nil, err
+	}
+
+	queries := remoteReadRequest.GetQueries()
+
+	for i, query := range queries {
+		params.Add(fmt.Sprintf("%d_start", i), fmt.Sprintf("%d", query.GetStartTimestampMs()))
+		params.Add(fmt.Sprintf("%d_end", i), fmt.Sprintf("%d", query.GetEndTimestampMs()))
```

Comment:
I would find `start[0]` more obvious, but maybe you're aiming for easier sorting?

Reply:
neither would help with sorting without padding, I just wrote the first idea, will switch to `start[0]`

Reply:
For now I'm sticking to param_start_0  format. 

The two main use-cases in my mind:
- search for a specific query => doesn't really matter where it is in the log line
- make a histogram of queries => I'm assuming that queries are not randomly placed in queries (?) so making a histogram on the first query can give a pretty good result, if not, you can keep adding more to the group by.

I've considered these:
- single `param_query=<json>`  that has everything. However that would be JSON array and [getting](https://grafana.com/docs/loki/latest/query/log_queries/#json) to a single element requires some extra steps which doesn't seem worth it to me.
- single `param_query=<logfmt>` that has everything. Probably quoting hell, unreadable to human.
- multiple `param_start_#="..."` with dedicated params for each field, except hints which shall be a JSON for taking advantage of omitempty mechanism.